db/src/tx_observer.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::sync::{
+    Arc,
+};
+use std::thread;
+
+use indexmap::{
+    IndexMap,
+};
+
+use types::{
+    AttributeSet,
+    TxReport,
+};
+
+pub struct TxObserver {
+    notify_fn: Arc<Option<Box<Fn(String, Vec<TxReport>) + Send + Sync>>>,
+    attributes: AttributeSet,
+}
+
+impl TxObserver {
+    pub fn new<F>(attributes: AttributeSet, notify_fn: F) -> TxObserver where F: Fn(String, Vec<TxReport>) + 'static + Send + Sync {
+        TxObserver {
+            notify_fn: Arc::new(Some(Box::new(notify_fn))),
+            attributes,
+        }
+    }
+
+    pub fn applicable_reports(&self, reports: &Vec<TxReport>) -> Vec<TxReport> {
+        reports.into_iter().filter_map( |report| {
+            if self.attributes.intersection(&report.changeset).next().is_some(){
+                Some(report.clone())
```

Comment:
Can we use `&TxReport`, or at least `Arc<TxReport>` here? Seems like there could be a lot of cloning going on.